cmd/crd-linter/exceptions/exceptions.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exceptions
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+)
+
+type ExceptionList struct {
+	Exceptions []Exception
+}
+
+type Exception struct {
+	filename, crdName, linter, fieldRef string
+}
+
+func (e Exception) String() string {
+	return fmt.Sprintf("%s|%s|%s|%s", e.filename, e.crdName, e.linter, e.fieldRef)
+}
+
+func (l *ExceptionList) IsExcepted(filename, crdName, linter, fieldRef string) bool {
+	toFind := Exception{
+		filename: filename,
+		crdName:  crdName,
+		linter:   linter,
+		fieldRef: fieldRef,
+	}
+	for _, e := range l.Exceptions {
+		if e == toFind {
+			return true
+		}
+	}
+	return false
+}
+
+func (l *ExceptionList) Add(filename, crdName, linter, fieldRef string) {
+	if l.IsExcepted(filename, crdName, linter, fieldRef) {
+		return
+	}
+	l.Exceptions = append(l.Exceptions, Exception{
+		filename: filename,
+		crdName:  crdName,
+		linter:   linter,
+		fieldRef: fieldRef,
+	})
+}
+
+func (l *ExceptionList) String() string {
+	buf := &strings.Builder{}
+	for _, e := range l.Exceptions {
+		buf.WriteString(fmt.Sprintf("%s\n", e.String()))
+	}
+	return buf.String()
+}
+
+func (l *ExceptionList) Size() int {
+	return len(l.Exceptions)
+}
+
+func (l *ExceptionList) WriteToFile(path string) error {
+	return ioutil.WriteFile(path, []byte(l.String()), 0644)
+}
+
+func New() *ExceptionList {
+	return &ExceptionList{Exceptions: make([]Exception, 0)}
+}
+
+// LoadFromFile will load a pipe (|) separated list of Exceptions from the
```

Comment:
This doc appears a bit confusing to me. The exception list seems to be separated by newline; one exception per line. But one exception seems to be serialized in a pipe-separated format.

Have you considered other file formats? The suggested format is very compact, but the semantics are not apparent when looking at a file.

Reply:
👍 the format here wasn't so much chosen to be compact, but rather just to have each entry on a single line to make it clear to see how many violations there are. An example line:

```
path/to/the/file/jaegertracing.io_jaegers_crd.yaml|jaegers.jaegertracing.io|MaxLengthStrings|spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.collector.properties.volumes.items.properties.flexVolume.properties.fsType.maxLength is not specified on a field of type 'string'
```

Happy to adjust this format however you feel makes it clearer - what sort of format do you think would be better/what do you have in mind?

Reply:
@munnerz Sorry for the delay. You have reasonable arguments for selecting the chosen file format, and I think it should be ok. Only downside I see, is that it is not very human readable. The format must remain stable over time, so we should decide if human readability is important now. I'll guess an IDE could be a consumer of these files. Do we have any reference or common practice to refer to?